src/server/helpers/wikimedia.ts

```
@@ -0,0 +1,166 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+import type {WikipediaArticle, WikipediaPageExtract} from '../../common/helpers/wikimedia';
+import {cacheJSON, getCachedJSON} from '../../common/helpers/cache';
+import {toLower, uniq} from 'lodash';
+import {hoursToSeconds} from 'date-fns';
+import request from 'superagent';
+import {userAgent} from '../info';
+
+
+type WikidataSiteLink = {
+
+	/**
+	 * Abbreviation for the Wikimedia project/site/domain.
+	 * Format: language code + project suffix (e.g. `enwiki` for the English Wikipedia)
+	 */
+	site: string,
+
+	/** Title of the specific page. */
+	title: string,
+
+	/** Wikidata IDs of badges which the page has (e.g. `Q17437798` for good articles). */
+	badges: string[],
+};
+
+type WikidataSiteLinksResult = {
+	entities: Record<string, {
+		id: string,
+		sitelinks: Record<string, WikidataSiteLink>,
+		type: 'item',
+	}>,
+	success: boolean,
+};
+
+// incomplete, only the parts we need
+type WikipediaExtractResult = {
+	query: {
+		pages: WikipediaPageExtract[],
+	},
+};
+
+
+/** Maximum age of cached results in seconds. */
+const cacheMaxAge = {
+	articles: hoursToSeconds(24 * 7),
+	extract: hoursToSeconds(24 * 3)
+};
+
+/**
+ * Fetches a list of Wikipedia articles in all available languages for the given Wikidata item.
+ * @param {string} wikidataId - Wikidata item ID.
+ */
+export async function getAvailableWikipediaArticles(wikidataId: string, {
+	forceCache = false
+} = {}): Promise<WikipediaArticle[]> {
+	const cacheKey = `wiki:articles:${wikidataId}`;
+	const cachedArticles = await getCachedJSON<WikipediaArticle[]>(cacheKey);
+
+	if (cachedArticles || forceCache) {
+		return cachedArticles || [];
+	}
+
+	const apiUrl = new URL('https://www.wikidata.org/w/api.php');
+	apiUrl.search = new URLSearchParams({
+		action: 'wbgetentities',
+		format: 'json',
+		ids: wikidataId,
+		props: 'sitelinks'
+	}).toString();
+
+	const response = await request.get(apiUrl.href)
```

This can fail; like my other comments, I'd like to see it wrapped in a try…catch block.

---

In theory I could already catch these errors inside the Wikimedia helper functions but at this stage it is difficult to decide how they should be treated. There's even one case where I considered it to be better to throw an error myself:
https://github.com/metabrainz/bookbrainz-site/blob/b0759410469e0f5cd5085cfc7a5612d9b618dd80/src/server/helpers/wikimedia.ts#L92-L94
So my intention was to let these errors propagate until they reach the middleware layer where they are actionable and catch them there.
I had already done this for the JSON route but forgot to handle this in the (newer) pre-loader middleware as well, which I did now in 307876a.

---

Great, that's totally fine now.